materialize-snowflake/stream.go

```diff
@@ -0,0 +1,536 @@
+		return fmt.Errorf("initializing bucket: %w", err)
+	}
+
+	fName := sm.getNextFileName(time.Now(), fmt.Sprintf("%s_%d", sm.prefix, sm.deploymentId))
+	w := sm.bucket.NewWriter(ctx, path.Join(sm.bucketPath, string(fName)), sm.objMetadata())
+
+	ts := sm.tableStreams[binding]
+	bdecWriter, err := newBdecWriter(w, ts.mappedColumns, ts.channel.TableColumns, ts.channel.EncryptionKey, fName)
+	if err != nil {
+		return fmt.Errorf("newBdecWriter: %w", err)
+	}
+	sm.bdecWriter = bdecWriter
+
+	return nil
+}
+
+func (sm *streamManager) objMetadata() blob.WriterOption {
+	return blob.WithObjectMetadata(map[string]string{
+		"ingestclientname": fmt.Sprintf("%s_EstuaryFlow", sm.tenant),
+		"ingestclientkey":  sm.prefix,
+	})
+}
+
+func (sm *streamManager) flush(baseToken string) (map[int][]*blobMetadata, error) {
+	if sm.bdecWriter != nil {
+		if err := sm.finishBlob(); err != nil {
+			return nil, fmt.Errorf("finishBlob: %w", err)
+		}
+	}
+
+	out := make(map[int][]*blobMetadata)
+	for binding, trackedBlobs := range sm.blobStats {
+		for idx, trackedBlob := range trackedBlobs {
+			out[binding] = append(out[binding], generateBlobMetadata(
+				trackedBlob,
+				sm.tableStreams[binding].channel,
+				blobToken(baseToken, idx)),
+			)
+		}
+	}
+	maps.Clear(sm.blobStats)
+
+	return out, nil
+}
+
+// write registers a series of blobs to the table. All the blobs must be for the
+// same table.
+//
+// Exactly-once semantics are achieved using the offset tokens for the blobs:
+// The offset token consists of the "base" token which is a string, and for each
+// blob it is appended with a sequential counter, starting at 0. So if there are
+// multiple blobs to append, they are organized with offset tokens like
+// "basetoken:0", "basetoken:1", "basetoken2:" etc.
+//
+// Blobs are registered in this order, and it's possible that only some of the
+// blobs get registered in a single Acknowledge before the connector fails for
+// some reason, or all of the blobs get registered but the Acknowledge response
+// is not persisted to the runtime before the connector restarts. The Snowpipe
+// channel itself persists the last registered token, and this allows us to
+// filter out blobs that had previously been registered and don't need
+// registered again on a re-attempt of an Acknowledge.
+func (sm *streamManager) write(ctx context.Context, blobs []*blobMetadata) error {
+	if err := validWriteBlobs(blobs); err != nil {
+		return fmt.Errorf("validWriteBlobs: %w", err)
+	}
+
+	var schema = blobs[0].Chunks[0].Schema
+	var table = blobs[0].Chunks[0].Table
+	var channelName = blobs[0].Chunks[0].Channels[0].Channel
+	var thisChannel *channel
+	for _, v := range sm.tableStreams {
+		matches := v.channel.Schema == schema && v.channel.Table == table && v.channel.Channel == channelName
+		if matches && thisChannel != nil {
+			return fmt.Errorf("internal error: found duplicate duplicate channel %s in tableStreams", channelName)
```

Comment:
```suggestion
			return fmt.Errorf("internal error: found duplicate channel %s in tableStreams", channelName)
```

Reply:
Some kind of subliminal force at work here where I typed the word "duplicate" twice 😃